### PR TITLE
set_user.dest_role_whitelist for set_user('rolename')

### PR DESCRIPTION
**Pull request by scrummyin**
I'm working on a project that could benefit from one user being able to become another. However, I want to limit the list of roles that anyone who has access to set_user can become. I added the guc value `set_user.dest_role_whitelist` to handle this and thought since it is a non-breaking change that I would create a pull request. 

Please let me know if it is acceptable, and if you would like more work done on it. Also, I included changes to the readme that I thought would be helpful.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -149,6 +155,15 @@ to ensure there are no other PostgreSQL roles existing which are both superuser
...
 
+#### Whitelist rules and logic
+
+For configuration options that end in `_whitelist` the following logic applies.
+
+* only roles explicitly listed or included by a group that is explicitly
+  listed (e.g. `'+admin'`) are allowed
+* if the whitelist is set to the empty set `''` all roles are denied
+* the default is `'*'` which means all roles are allowed
+
```
Comment (mpalmi) [inline README.md]: I like this in theory, but the whitelists have different meanings in this particular instance, since `set_user.superuser_whitelist` is for allowed sources and `set_user.dest_role_whitelist` is for allowed targets. Because of this, "allowed" and "denied" are sort of ambiguous.

We may also want to consider a different name for the whitelists, to distinguish between them. Maybe something like `set_user.superuser_source_whitelist` and `set_user.nonsuperuser_target_whitelist`?

File: set_user.c [hunk 2]
```diff
@@ -348,7 +349,11 @@ set_user(PG_FUNCTION_ARGS)
 						(errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
 						 errmsg("switching to superuser not allowed"),
 						 errhint("Add current user to set_user.superuser_whitelist.")));
-		}
+		} else if(!check_user_whitelist(NewUserId, Dest_role_whitelist))
```
Comment (mpalmi) [inline set_user.c]: style nit: `else if` should be on the next line:
```
}
else if ...
{
```

File: README.md [hunk 3]
```diff
@@ -63,13 +68,14 @@ Specifically, when an allowed user executes `set_user('rolename')` or
...
-permission on `set_user_u()` can escalate to superuser. The default value of
-`set_user.superuser_whitelist` is `'*'`.
+superuser. Additionally, all rules in [whitelist](#whitelist-rules-and-logic)
+apply to `set_user.superuser_whitelist` and `set_user_u('rolename')` with respect
+to the `rolename` argument.
+
+User's calling `set_user('rolename')` can only transition to roles listed or
+included in `set_user.dest_role_whitelist` (defaults to all roles). Additionally
+the [Whitelist logic](#whitelist-rules-and-logic) applies to `current_user` when
+`set_user()` is invoked.
```
Comment (mpalmi) [inline README.md]: Since we're going to revise this paragraph, maybe something like this instead:

> The following rules govern escalation to superuser via the `set_user_u('rolename')` function:
>
> * `current_user` must be `GRANT`ed `EXECUTE ON FUNCTION set_user_u('rolename')` OR `current_user` must be the OWNER of the `set_user_u` function OR `current_user` must be a superuser.
> * `current_user` must be listed in `set_user.superuser_whitelist` OR `current_user` must belong to a group that is listed in `set_user.superuser_whitelist` (e.g. `'+admin'`)
> * If `set_user.superuser_whitelist` is the empty set , `''`, superuser escalation is blocked for all users.
> * If `set_user.superuser_whitelist` is the wildcard character, `'*'`, all users with `EXECUTE` permission on `set_user_u('rolename')` can escalate to superuser.
> * If `set_user.superuser_whitelist` is not specified, the value defaults to the wildcard character, `'*'`.
>
> The following rules govern non-superuser role transitions through use of the `set_user('rolename')` function:
>
> * `current_user` must be `GRANT`ed `EXECUTE ON FUNCTION set_user('rolename')` OR `current_user` must be the OWNER of the `set_user` function OR `current_user` must be a superuser.
> * The destination `rolename` must be listed in `set_user.dest_role_whitelist` OR the destination `rolename` must belong to a group that is listed in `set_user.dest_role_whitelist` (e.g. `'+client'`)
> * If `set_user.dest_role_whitelist` is the empty set , `''`, `set_user` transitioning to non-superusers is blocked for all users.
> * If `set_user.dest_role_whitelist` is the wildcard character, `'*'`, all users with `EXECUTE` permission on `set_user('rolename')` can transition to any other non-superuser role.
> * If `set_user.dest_user_whitelist` is not specified, the value defaults to the wildcard character, `'*'`.

**Review by mpalmi — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (scrummyin) [reply README.md]: I don't think the whitelist have different meanings in these cases, and I think being able to re-use the check_user_whitelist function illistrates that. The difference is in the target of the whitelist current_user vs the rolename argument and maybe wording/naming can fix that.

That said, I completely agree with you on a name change. My original hesitation on it was that I did not want to rename `set_user.superuser_whitelist` as I thought it make reading GUC variables confusing. I was unsure if the setting should reading from `set_user.superuser_source_whitelist` and then `set_user.superuser_whitelist`, in order to keep compatiable with current setups.

In summary, I will rename `set_user.dest_role_whitelist` to `set_user.nonsuperuser_target_whitelist` and `*Dest_role_whitelist` to `*Nonsuperuser_Target_Whitelist`, as those names have better meaning and no one is using them yet. However, I'll wait for you feedback on renaming  `set_user.superuser_whitelist` and how to handle reading from the guc for it's value. The other question is while I'm working on it would you like the C variables of `*SU_Whitelist` renamed, and if so what to?

File: set_user.c (hunk 2 above)
Comment (scrummyin) [reply set_user.c]: Will do thanks for the catch on that one.

File: README.md (hunk 3 above)
Comment (scrummyin) [reply README.md]: Alright, I'll make these changes, but I'm going to wait on the feedback from above since if we are going to change the names, I'd rather only have to edit this section once.

Comment (mpalmi) [reply README.md]: The differences between these two paragraphs illustrate the ambiguity I referenced above. I think I'd prefer to leave the logic explanations split out, since we can be more specific about what the whitelist does for each feature.

Comment (scrummyin) [reply README.md]: Oh sorry misunderstanding here, I will keep it split. I meant if I re-write it now I would have to edit it again later if we rename `dest_user_whitelist` and `superuser_whitelist`. So, I want to add that whole section pretty much as is and make the variable name changes all in one go.

Comment (scrummyin) [reply README.md]: `as is` == `as you wrote it above`

File: README.md (hunk 1 above)
Comment (mpalmi) [reply README.md]: The re-use of `check_user_whitelist` is an implementation detail. Though it's nice that we can do that, it really has no bearing on the end-user that this doc is geared toward, IMO.

Good call re: `superuser_whitelist`. Maybe we leave that the same for now and move forward with the `dest_role` renaming. We may be able to do better than `nonsuperuser_target_whitelist`. Perhaps we should discuss the name out-of-band?

**Comment by scrummyin** [comment]
Alright, I think I have a more updated set of changes. Let me know if there is anything more I can do.

File: README.md [hunk 4]
```diff
@@ -134,10 +139,10 @@ Alternatively, transitions can be made to superusers through use of
...
 ```
 
-**Note:** Superuser escalation is only allowed for the roles listed in
-`set_user.superuser_whitelist`. If the whitelist is equal to `'*'`, all roles
-that have been granted `EXECUTE` on `set_user_u` can escalate to superuser.
-This is the default setting of `set_user.superuser_whitelist`.
+**Note:** See rules in [Superuser Whitelist](#set_usersuperuser_whitelist-rules-and-logic)
+for logic around calling set_user_u. See
+[Nosuperuser Whitelist](#set_usernosuperuser_target_whitelist-rules-and-logic) for
+reference logic around calling set_user.
```
Comment (mpalmi) [inline README.md]: `set_user`

File: README.md [hunk 5]
```diff
@@ -149,6 +154,44 @@ to ensure there are no other PostgreSQL roles existing which are both superuser
...
+* If `set_user.superuser_whitelist` is the wildcard character, `'*'`, all users
+  with `EXECUTE` permission on `set_user_u(text)` can escalate to superuser.
+* If `set_user.superuser_whitelist` is not specified, the value defaults to the
+  wildcard character, `'*'`.
+
+#### `set_user.nosuperuser_target_whitelist` Rules and Logic
+
+The following rules govern non-superuser role transitions through use of
+  `set_user(text)` or `set_user(text, text)` function (for simplicity
+  only `set_user(text)` is used):
```
Comment (mpalmi) [inline README.md]: > for simplicity, only `set_user(text)` is referenced below

File: README.md [hunk 6]
```diff
@@ -149,6 +154,44 @@ to ensure there are no other PostgreSQL roles existing which are both superuser
...
+
+* `current_user` must be `GRANT`ed `EXECUTE ON FUNCTION set_user(text)` OR
+  `current_user` must be the `OWNER` of the `set_user(text)` function OR
+  `current_user` must be a superuser.
+* The destination rolename must be listed in `set_user.nosuperuser_target_whitelist`
+  OR the destination rolename must belong to a group that is listed in
+  `set_user.nosuperuser_target_whitelist` (e.g. `'+client'`)
+* If `set_user.nosuperuser_target_whitelist` is the empty set , `''`, set_user
```
Comment (mpalmi) [inline README.md]: `set_user`

Comment (scrummyin) [reply README.md]: how about `set_user(text)`?

File: README.md [hunk 7]
```diff
@@ -149,6 +154,44 @@ to ensure there are no other PostgreSQL roles existing which are both superuser
...
+  transitioning to non-superusers is blocked for all users.
```
Comment (mpalmi) [inline README.md]: s/transitioning/transitions/

File: README.md (hunk 6 above)
Comment (mpalmi) [reply README.md]: :+1:

File: README.md [hunk 8]
```diff
@@ -149,6 +154,44 @@ to ensure there are no other PostgreSQL roles existing which are both superuser
...
+#### `set_user.nosuperuser_target_whitelist` Rules and Logic
+
+The following rules govern non-superuser role transitions through use of
+  `set_user(text)` or `set_user(text, text)` function (for simplicity
+  only `set_user(text)` is used):
+
+* `current_user` must be `GRANT`ed `EXECUTE ON FUNCTION set_user(text)` OR
+  `current_user` must be the `OWNER` of the `set_user(text)` function OR
+  `current_user` must be a superuser.
+* The destination rolename must be listed in `set_user.nosuperuser_target_whitelist`
```
Comment (mpalmi) [inline README.md]: s/destination/target/g

**Comment by mpalmi** [comment]
Looks good. Tested it out on my dev box for pg10/9.6.
